utils/misc.ml

```diff
@@ -181,15 +181,50 @@ let may_map = Stdlib.Option.map
 
 (* File functions *)
 
+let find_cache = Hashtbl.create 16
+let find_cache_cwd = ref (Sys.getcwd ())
+let reset_find_cache () =
+  Hashtbl.clear find_cache
+
+let really_exists dir name =
+  let dir = if dir = "" then Filename.current_dir_name else dir in
+  let () =
+    if Filename.is_relative dir then
+      let cwd = Sys.getcwd () in
+      if !find_cache_cwd <> cwd then begin
+        Hashtbl.clear find_cache;
+        find_cache_cwd := cwd
+      end
+  in
+  match Hashtbl.find find_cache dir with
+  | files -> Hashtbl.mem files name
+  | exception Not_found ->
+      try
+        let files = Sys.readdir dir in
+        let table = Hashtbl.create (Array.length files * 2) in
+        let add found file =
+          Hashtbl.add table file (); found || file = name
+        in
+        Hashtbl.add find_cache dir table;
+        Array.fold_left add false files
+      with _ ->
+        false
+
 let find_in_path path name =
   if not (Filename.is_implicit name) then
-    if Sys.file_exists name then name else raise Not_found
+    if really_exists (Filename.dirname name) (Filename.basename name) then
+      name
+    else
+      raise Not_found
   else begin
     let rec try_dir = function
       [] -> raise Not_found
     | dir::rem ->
-        let fullname = Filename.concat dir name in
-        if Sys.file_exists fullname then fullname else try_dir rem
+        let name = Filename.concat dir name in
+        if really_exists (Filename.dirname name) (Filename.basename name) then
```

Comment:
Sometimes you do `if really_exists (Filename.dirname name) (Filename.basename name) then name` and sometimes `if really_exists dir name then Filename.concat dir name`. I find it a bit hard to follow, why not just pass a single filename to `really_exists`?